### PR TITLE
New filters for satellites

### PR DESCRIPTION
**Pull request by alex-w**
Added 2 new filters for artificial satellites

### Description
I've added two new filters for artificial satellites - "[all user defined]" and "[outdated TLE]". The filter "[all user defined]" was introduced many years ago, but it was not visible for users. By the fact this filter will display satellites, which have activated option "Do not update" in the GUI, e.g. user defined satellites (handmade TLE) or satellites with forbidden updating the TLE. The filter "[outdated TLE]" is newly introduced filter and I added him to get a list of satellites, which have outdated epoch of TLE. By the fact I mark epoch of TLE as oudated if it have age 30 days ago or more. For example satellite with TLE epoch from year 2019 is really oudated IMHO.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
@JAY-Githb, please review this pull request too

**Comment by gzotti** [comment]
Hmm, I usually have disabled this plugin and am not really very familiar... But I will try to review the code.

File: plugins/Satellites/src/Satellite.cpp [hunk 1]
```diff
@@ -864,6 +865,8 @@ SatFlags Satellite::getFlags() const
 		flags |= SatHEO;
 	if (eccentricity < 0.25 && (inclination>=25. && inclination<=180.) && (orbitalPeriod>=1100. && orbitalPeriod<=2000.))
 		flags |= SatHGSO;
+	if (StelApp::getInstance().getCore()->getJD() - tleEpochJD > 30.) // TLE epoch is over 30 days old
```
Comment (gzotti) [inline plugins/Satellites/src/Satellite.cpp:868]: Not sure: is current simulation time critical, or rather real-world time? (getJDfromSystem() ?) And: are TLE also outdated if I go 4 months back? in the latter case, at least test abs(JD-tleEpochJD)

Comment (alex-w) [reply plugins/Satellites/src/Satellite.cpp:868]: Simulation time is critical IMHO. TLE is valid on time range 72 hours (probably up to 100 hours with low accuracy). I'm use multiplicator 10 for the 72 hours (3 days * 10 = 30 days), but yes, `abs()` will be good here.

**Comment by gzotti** [comment]
OK, else this should at least not break anything (famous last words...), go ahead.

**Comment by alex-w** [comment]
> OK, else this should at least not break anything (famous last words...), go ahead.

I want to see comments from @JAY-Githb as old-school satellites observer :)

**Comment by JAY-Githb** [comment]
Hello.
I'm not THAT old. I've only been observing since June 1958  ;)   (Sputnik 3 rocket).  
If I knew programing, I would be glad to help.  You are doing a fabulous job with enhancing Stellarium!   I'm going to look for that update.   
Thank you,

**Comment by alex-w** [comment]
> I'm not THAT old. I've only been observing since June 1958 ;) (Sputnik 3 rocket).

:)

> If I knew programing, I would be glad to help. You are doing a fabulous job with enhancing Stellarium! I'm going to look for that update.

Hmm... What about the answer for question: how acceptable mark a TLE as outdated if epoch of TLE at least 30 days old?

**Comment by gzotti** [comment]
I think this is a good idea.

**Comment by JAY-Githb** [comment]
Hello.
Marking an old TLE is a good idea as long as the age can be chosen. Many objects in high orbits can still be found with 60 day old TLEs. I used to get predictions by regular mail that way. With internet now, most elements are updated regularly. The only objects likely to have older TLE are classified. But as you point out, TLEs from months ago are useless so removing them from predictions is a helpful idea.

Jay Respler

**Comment by alex-w** [comment]
**Important note:** this code is used for get a filtered list of satellites in GUI of the plugin. To removing or not the satellite with outdated TLE should be a human. The option to define the age of TLE's epoch is acceptable and it can be implemented by adding config option + GUI element for it.

**Comment by alex-w** [comment]
@gzotti could you check my bad English? If PR is OK then please merge it.

**Comment by JAY-Githb** [comment]
Alex, you are right the filter should just not predict for old TLE.  Let the user determine which TLE to remove.

Jay

**Comment by alex-w** [comment]
@JAY-Githb please check the fresh release

**Comment by JAY-Githb** [comment]
HI.
I am looking at it and have some questions. I don't see an explanation in the user manual.

> Added also search descriptions in Satellites plugin (GH: #1940)

How does this work differently than before?

> Added new filter for satellites (GH: #2124)

I see that setting, but how does it activate?  I thought it automatically would not display satellites if the TLE was older than 30 days. I checked a couple of objects and they display normally.

> Changed Satellites plugin: improve guessing GALILEO satellites

Sorry, I don't understand what that  does  or how to use it.

Best wishes, 
Jay Respler

**Comment by alex-w** [comment]
> > Added also search descriptions in Satellites plugin (GH: #1940)
> 
> How does this work differently than before?

Yes, no search by field "description" before. Now:
sat-station

> > Added new filter for satellites (GH: #2124)
> 
> I see that setting, but how does it activate? I thought it automatically would not display satellites if the TLE was older than 30 days. I checked a couple of objects and they display normally.

No. This is just a filter for satellites (as I saw before):
sat-filter

> > Changed Satellites plugin: improve guessing GALILEO satellites
> 
> Sorry, I don't understand what that does or how to use it.

Adding satellites into group "galileo"

**Comment by JAY-Githb** [comment]
Ok, it took some experimenting, but I understand how they work now.
Thank you.
I don't want to interfere with the fine work you do, but if  you could use my help writing a description for the  User Guide of how to use these features, I am willing to do so.

Jay

**Comment by alex-w** [comment]
> I don't want to interfere with the fine work you do, but if you could use my help writing a description for the User Guide of how to use these features, I am willing to do so.

If you can help us doing Stellarium better (by the code, documenting, bug reporting, etc.) then welcome! :)

**Comment by gzotti** [comment]
Yes, an update to the documentation in the User Guide written by somebody who is actively using this plugin would be very welcome. If you are familiar with the LaTeX typesetting system, you can directly develop the chapter. Else, copy/paste the chapter and submit some other document format (Word, OpenOffice, ...) which we can re-format. 
If you feel some buttons or button tooltips are missing/ill-described, also please tell us.

**Comment by JAY-Githb** [comment]
Hi Alex and Georg.
I can't do coding or Latex but I can write descriptions of how to use features and send you the text.  It was hard to use Stellarium when I first started because  many features had little or no explanation of how to use them.  It took me a long time experimenting to figure them out.  Now that I am familiar with it, I think Stellarium is the best mapping program. I don't remember which features were not well documented but as I come across them, I can let you know.
To start, I'll review the plug-in changes that were just made.
Should I post notes here or send them direct to your personal email?
You can reach me at   JRespler@superlink.net

Alex, are you a member of Seesat-L?  It is the group for satellite observers.

Best wishes,
Jay

**Comment by alex-w** [comment]
> Should I post notes here or send them direct to your personal email?

You may use maillist address (stellarium@googlegroups.com), discussions & issues here, maybe pull requests later and, of course, personal emails.

> Alex, are you a member of Seesat-L? It is the group for satellite observers.

Yes, of course. For example I used Anthony Mallama works for prediction magnitude of Starlink satellites.

**Comment by JAY-Githb** [comment]
In his article in Sky & Telescope magazine about Starlink, Anthony stated that his work was based on observations made by two people. His and mine.  I'm glad to see that some of my observations are being used in Stellarium.
If you don't get Sky & Telescope, I can send you the article.

Looking at the User Guide, I see that most of the Satellite plug-in discusses TLEs and sources.  There is almost no description of what features are available in Stellarium and how to use them.  Someone taking a quick look at the User Guide would not realize how useful Stellarium could be.  
If you and Georg think it would help, after I review the latest additions, I could write up a description of how to use the Satellite Configuration features.

**Comment by gzotti** [comment]
The core of the satellites plugin was written about a decade ago by developers who unfortunately have left the project and did not write a full documentation. Over the years some more switches and options were added as required. Obvious for the developers, mysteries for the newbie. I did occasionally prepare for Iridium flashes, but else am not active here so I cannot write the pretty details.  (I have once made a time-lapse of geostationaries identifiable with Stellarium going into earth shadow, though.) If you feel inclined, please rewrite chapter 13.9 with the grade of detail you think best helps new and seasoned satellite observers. (But maybe not more than about 10 pages :-)

**Comment by JAY-Githb** [comment]
Ha!   You can be sure it won't be more than 10 pages. I never liked writing so I would keep things short, but clear enough to understand. 
As Iridium flares are no longer occurring, predictions for them can be removed from the program.

Is  the time-lapse of geostationaries shown somewhere in Stellarium?

General feature request: All Windows have black text on dark gray background making them hard to read.  Can either text or background be make lighter to improve the contrast?

Thanks.

**Comment by gzotti** [comment]
Good. Concise and complete is fine :-)
We have removed the Iridiums a year ago or so. 
Timelapse is unpublished, but I could dig out my video for you.

**Comment by JAY-Githb** [comment]
If the timelapse is not already in Stellarium,  don't bother digging it out.  I can already simulate that with the nice features already in the program.

**Comment by JAY-Githb** [comment]
*    Added also search descriptions in Satellites plugin (GH: #1940)

I realize someone requested this, but it seems to have little use.  Most people won't take the time to add many comments in the Description.
Much more useful would be able to Search for International Designation.  
For instance a search for 98067 would bring up ISS and all the other objects put into orbit from there.
Search for 21115 would show all Starlinks, retaining rods, rideshare Global satellites and the mounting adapter.
Search for debris from an explosion could show the dispersal over time.

Then in addition to sorting by name now, ability to sort by Norad # or International ID would be useful.  
Either automatically sort by search method used, or Maybe a check box to allow user to choose which sort method to use.

**Comment by JAY-Githb** [comment]
* Added new filter for satellites (GH: #2124)

This helps avoid predictions for old satellites. That is good.

We had discussed a filter that would be even more useful.  Just as on abject in shadow is not displayed, a filter that shows objects only when within a minimum/maximum range  would help un-clutter the screen.  It is difficult to pick out potential objects to view when they all look the same no matter what the distance is.

**Comment by gzotti** [comment]
Well, I did it... compare https://1drv.ms/v/s!AuJ_d8wlPcFpk2SnaigLCvdlhm26?e=CxaIIt against https://1drv.ms/v/s!AuJ_d8wlPcFpk2UTZ0rxCQ1iRTX-?e=YPcZQT. The series covers about 3.5 hours. The real thing is still better :-)

**Comment by alex-w** [comment]
> I realize someone requested this, but it seems to have little use. Most people won't take the time to add many comments in the Description.

But someone may use it :)

> Much more useful would be able to Search for International Designation.

OK, I may add able to Search for International Designation as for NORAD number (COSPAR XXXX-XXXX or just XXXX-XXXX)

> For instance a search for 98067 would bring up ISS and all the other objects put into orbit from there. Search for 21115 would show all Starlinks, retaining rods, rideshare Global satellites and the mounting adapter.

Hmm... But way not use the full designation like 1998-067A?

> Then in addition to sorting by name now, ability to sort by Norad # or International ID would be useful. Either automatically sort by search method used, or Maybe a check box to allow user to choose which sort method to use.

Where is? The search by NORAD number is implemented in Search Tool and in GUI/Satellites tab of plugin.

**Comment by alex-w** [comment]
> We had discussed a filter that would be even more useful. Just as on abject in shadow is not displayed, a filter that shows objects only when within a minimum/maximum range would help un-clutter the screen. It is difficult to pick out potential objects to view when they all look the same no matter what the distance is.

**Important note:** in the pull request was added 2 new filters for GUI/Satellites tab of plugin for get a filtered list of satellites in the database of all loaded satellites. But in other discussion (issue #2060) for distance range filter you want to get a filter for visibility of satellites. This is different filters and the second one is difficult to implement (it should support 2 modes + problem for rendering orbits). Or you want to get a filter to get a list of satellites to change the visibility by hands (checking the distance range by apogee/perigee is not very difficult to implement)?

**Comment by JAY-Githb** [comment]
>     Much more useful would be able to Search for International Designation.

OK, I may add able to Search for International Designation as for NORAD number (COSPAR XXXX-XXXX or just XXXX-XXXX)

    For instance a search for 98067 would bring up ISS and all the other objects put into orbit from there. Search for 21115 would show all Starlinks, retaining rods, rideshare Global satellites and the mounting adapter.

> Hmm... But way not use the full designation like 1998-067A?

I said 98067 because that is the way it is listed in TLE.  However, as you suggest, having user enter the entire designation like 1998-067 might make it easier to pick out the correct value from the TLE.
 I didn't add the  'A'  because we're looking for All objects associated with that launch.  If adding "A' makes for a better search for all the other objects, then certainly do that.

**Comment by JAY-Githb** [comment]
>   Then in addition to sorting by name now, ability to sort by Norad # or International ID would be useful. Either automatically sort by search method used, or Maybe a check box to allow user to choose which sort method to use.

> Where is? The search by NORAD number is implemented in Search Tool and in GUI/Satellites tab of plugin.

Yes, we can search by  NORAD but the sorted output is shown alphabetically. 

For example, search for K-125 gives me these objects:
STARLINK- 1254
STARLINK- 1255
STARLINK- 1256
STARLINK- 1257
STARLINK- 1258
STARLINK- 1259

But if the output sorts by NORAD you get
STARLINK-1254  (45216)
STARLINK-1258  (45381)
STARLINK-1255  (45399)
STARLINK-1256  (45401)
STARLINK-1257  (45402)
STARLINK-1259  (45403)

Sorted by  International ID
 STARLINK-1254   (20-12AQ)
 STARLINK-1255   (20-19AR)
 STARLINK-1256   (20-19AT)
 STARLINK-1257   (20-19AU)
 STARLINK-1259   (20-19AV)
 STARLINK-1258   (20-19X )

So it's the same search results, but each displays differently based on the sort method chosen.

It's a nice feature but can certainly wait till next year       ;)

**Comment by JAY-Githb** [comment]
> > We had discussed a filter that would be even more useful. Just as on abject in shadow is not displayed, a filter that shows objects only when within a minimum/maximum range would help un-clutter the screen. It is difficult to pick out potential objects to view when they all look the same no matter what the distance is.
> 
> **Important note:** in the pull request was added 2 new filters for GUI/Satellites tab of plugin for get a filtered list of satellites in the database of all loaded satellites. But in other discussion (issue #2060) for distance range filter you want to get a filter for visibility of satellites. This is different filters and the second one is difficult to implement (it should support 2 modes + problem for rendering orbits). Or you want to get a filter to get a list of satellites to change the visibility by hands (checking the distance range by apogee/perigee is not very difficult to implement)?

Sorry, I'm not sure I understand the question.  If  ' visibility of satellites' means Magnitude, I dropped that completely.  Let me know if I"m mistaken about visibility.
I had originally mentioned a magnitude filter.  Georg had suggested a much better filter, so completely ignore magnitude.
Georg suggested Range. That is an excellent idea.   Since that is already calculated for each object, just hide the satellite if above/below the chosen amount.  If both values are too difficult, then just use a maximum range as the filter.
 Apogee/perigee  can also be ignored because that has no relation to a particular observer.
For instance Hubble Space Telescope  is in a circular orbit.  At low latitudes it can be directly overhead at a small range.  At high latitudes, it is very low on the horizon. Same altitude, but a much different range. 
That's why a range filter is better than an altitude filter.
No concern about magnitude at all.

**Pull request merged**